src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,248 @@
+  }
+
+  @Override
+  public void stop() {
+    this.executor.shutdown();
+
+    try {
+      this.executor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ex) {
+      throw new AssertionError(ex);
+    }
+
+    try {
+      reconciliationCache.leaveWorkerSet(workerId);
+    } catch (Exception ex) {
+      logger.warn("failed to remove from worker set", ex);
+    }
+  }
+
+  @Override
+  public void run() {
+    boolean accelerate  = false;
+    long    workerCount = 1;
+
+    try {
+      if (reconciliationCache.lockActiveWorker(workerId, CHUNK_INTERVAL)) {
+        try {
+          if (processChunk()) {
+            accelerate = reconciliationCache.isAccelerated();
+          }
+        } finally {
+          reconciliationCache.unlockActiveWorker(workerId);
+        }
+      }
+
+      if (!accelerate) {
+        workerCount = Math.max(reconciliationCache.getWorkerCount(WORKER_TTL_MS), 1);
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    if (!accelerate) {
+      scheduleWithJitter(CHUNK_INTERVAL * workerCount);
+    } else {
+      scheduleWithJitter(ACCELERATE_CHUNK_INTERVAL);
+    }
+  }
+
+  private void scheduleWithJitter(long delayMs) {
+    long randomJitterMs = (long) (random.nextDouble() * JITTER_MAX * delayMs);
+    try {
+      executor.schedule(this, delayMs + randomJitterMs, TimeUnit.MILLISECONDS);
+    } catch (RejectedExecutionException ex) {
+      logger.info("reconciler shutting down");
+    }
+  }
+
+  private boolean processChunk() {
+    Optional<String>               fromNumber          = reconciliationCache.getLastNumber();
+    int                            chunkSize           = (int) Math.max(getAccountCount() * PERIOD / CHUNK_INTERVAL, MINIMUM_CHUNK_SIZE);
+    DirectoryReconciliationRequest request             = readChunk(fromNumber, chunkSize);
+    Response                       sendChunkResponse   = sendChunk(request);
+    boolean                        sendChunkSuccessful = sendChunkResponse.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL;
+
+    if (sendChunkResponse.getStatus() == 404 || request.getToNumber() == null) {
+      reconciliationCache.clearAccelerate();
+    }
+
+    if (sendChunkSuccessful) {
+      reconciliationCache.setLastNumber(Optional.fromNullable(request.getToNumber()));
+    } else if (sendChunkResponse.getStatus() == 404) {
+      reconciliationCache.setLastNumber(Optional.absent());
+    }
+
+    return sendChunkSuccessful;
+  }
+
+  private DirectoryReconciliationRequest readChunk(Optional<String> fromNumber, int chunkSize) {
+    try (Timer.Context timer = readChunkTimer.time()) {
+      List<Account> accounts = accountsManager.getAllFrom(fromNumber, chunkSize);
+
+      List<String> numbers = new ArrayList<>(accounts.size());
+      for (Account account : accounts) {
+        if (account.isActive()) {
+          numbers.add(account.getNumber());
+        }
+      }
```

Comment:
You can do accounts.stream().filter(Account::isActive).map(Account::getNumber) these days if you want =)

Reply:
nice. i guess i'm confused, considering our whole guava vs. java Optional debacle, what modern java features i can actually use?

Reply:
AFAIK, we use Java 8. Java 8 provides both streams and optional. Seems like both are fair game.

Reply:
With the caveat that you all have to use the guava optional (used everywhere else in the current code base) until I can get the secretsender changes in on top of this change, which makes the jump to using the oracle optional class everywhere.

Reply:
fixed in 8ed0058f3703